### PR TITLE
sha: add command-line key/IV configuration and file input support

### PR DESCRIPTION
**Pull request by Akshay-Belsare**
This patch enhances the SHA example to support runtime key and IV configuration via command-line arguments, file input processing, and adds HMAC_SHA3 algorithm support.

The implementation allows users to provide cryptographic keys and IVs directly as hexadecimal strings on the command line, making it easy to test different keys without modifying source code. Strict validation ensures key and IV sizes match the algorithm requirements.

File input support enables processing any file type by reading the complete file contents. This is essential for real-world use cases.

Supported algorithms:
- HMAC: SHA1, SHA224, SHA256, SHA384, SHA512 (64-128 byte keys)
- HMAC_SHA3: SHA3_224, SHA3_256, SHA3_384, SHA3_512 (128 byte keys)
- CMAC: AES_CMAC (16 byte key + 16 byte IV)
- Digest: SHA1, SHA224, SHA256, SHA384, SHA512, SHA3_*, SHAKE*

All command-line arguments are optional with sensible defaults: HMAC_SHA256 algorithm, "test message" input, 0xa5 pattern keys, and 0x00 pattern IVs. This enables quick testing without creating input files.

Usage:
======

1. Quick test with all defaults: optee_example_sha

2. Test specific algorithm: optee_example_sha --algo HMAC_SHA3_256

3. Process file: optee_example_sha --algo SHA256 --input-file firmware.bin

4. HMAC with custom key (256 hex chars = 128 bytes): optee_example_sha --algo HMAC_SHA256 --input-file data.bin \ --key a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5...

5. AES_CMAC with key and IV (32 hex chars each = 16 bytes): optee_example_sha --algo AES_CMAC --input-file secure.dat \ --key c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3 \ --iv 00000000000000000000000000000000

### CODE REVIEW

File: sha/host/main.c [hunk 1]
```diff
@@ -50,18 +99,140 @@ static void terminate_tee_session(struct test_ctx *ctx)
...
+	}
+
+	if (byte_len > max_bytes) {
+		fprintf(stderr, "Error: Hex string too long (max %zu bytes)\n", max_bytes);
+		return -1;
+	}
+
+	for (size_t i = 0; i < byte_len; i++) {
+		unsigned int byte_val;
+		if (sscanf(hex_str + (i * 2), "%2x", &byte_val) != 1) {
```
Comment (jforissier) [inline sha/host/main.c]: This is a bit fragile, for instance "0g" would return 0 in `byte_val`. I suggest using `isxdigit()` on each character

File: sha/host/main.c [hunk 2]
```diff
@@ -50,18 +99,140 @@ static void terminate_tee_session(struct test_ctx *ctx)
...
+
+	*buffer = data;
+	*size = (size_t)file_size;
+	return 0;
+}
+
+static void print_usage(const char *progname)
 {
-	fprintf(stderr, "Usage: %s <string to process> [<algo>]\n\n", argv[0]);
+	fprintf(stderr, "Usage: %s [options]\n\n", progname);
```
Comment (jforissier) [inline sha/host/main.c:213]: The new syntax does not allow to provide a text message on the command line, I suggest adding `--input "string to process"` for easier compatibility

File: sha/host/main.c [hunk 3]
```diff
@@ -203,151 +374,248 @@ static TEEC_Result compare_hmac_sha(struct test_ctx *ctx, void *message,
...
+			if (algo_info->needs_iv && user_iv_len != algo_info->iv_size) {
+				fprintf(stderr, "Error: IV size mismatch. Expected %zu bytes, got %zu bytes for %s\n",
+				     algo_info->iv_size, user_iv_len, algo_info->name);
+				if (free_message && file_buffer)
+					free(file_buffer);
+				return -1;
+			}
+			printf("Using IV from command line (%zu bytes)\n", user_iv_len);
 		}
+
```
Comment (jforissier) [inline sha/host/main.c:536]: I suggest also checking:
```C
if (key_hex && !algo_info->needs_key) {
	fprintf(stderr, "Error: --key is not valid for %s\n", algo_info->name);
	return 1;
}

if (iv_hex && !algo_info->needs_iv) {
	fprintf(stderr, "Error: --iv is not valid for %s\n", algo_info->name);
	return 1;
}
```

**Review by jforissier — COMMENTED** [review]
Some small comments otherwise LGTM. Please add:

`Acked-by: Jerome Forissier <jerome.forissier@arm.com>`

File: sha/host/main.c (hunk 2 above)
Comment (amey-raghatate) [reply sha/host/main.c:213]: Hi @jforissier 
Thank you for the suggestion. I believe adding --input <string> would be redundant since the example already provides two ways to supply input data: --input-file for file-based input and a built-in default "test message" for quick testing. Adding a third input path would increase code complexity. Please let us know if you feel strongly otherwise.

Comment (jforissier) [reply sha/host/main.c:213]: This PR is breaking backwards compatibility, since the syntax has changed. I'm OK with that since the new syntax provides more flexibility. That being said, requiring users to update the command line is one thing ; requiring them to create a file, run the command, then cleanup is a bit much. Consider the following example:
```
optee_example_sha "Hello, World!"
```
In the general case it needs to be converted to:
```
  optee_example_sha_string() (
    tmp=$(mktemp) || exit 1
    trap 'rm -f "$tmp"' EXIT
    printf '%s' "$1" >"$tmp" || exit 1
    optee_example_sha --input-file "$tmp"
  )

  optee_example_sha_string 'Hello, World!'
```
...which is much more complicated than:
```
optee_example_sha --input "Hello, World!"
```

File: sha/host/main.c [hunk 4]
```diff
@@ -120,10 +120,12 @@ static int hex_to_bytes(const char *hex_str, uint8_t *bytes, size_t max_bytes, s
 
 	for (size_t i = 0; i < byte_len; i++) {
 		unsigned int byte_val;
-		if (sscanf(hex_str + (i * 2), "%2x", &byte_val) != 1) {
+		if (!isxdigit((unsigned char)hex_str[i * 2]) ||
+		    !isxdigit((unsigned char)hex_str[i * 2 + 1])) {
```
Comment (jforissier) [inline sha/host/main.c:124]: Needs `<ctype.h>`

File: sha/host/main.c [hunk 5]
```diff
@@ -203,151 +378,267 @@ static TEEC_Result compare_hmac_sha(struct test_ctx *ctx, void *message,
...
+
+	ctx.algo_num = algo_info->tee_algo_id;
+
+	/* Handle key and IV: command-line or defaults */
+	if (key_hex || iv_hex) {
+		/* Option 1: Load from command-line arguments */
+		if (key_hex) {
+			if (hex_to_bytes(key_hex, user_key, MAX_KEY_SIZE, &user_key_len) != 0) {
+				fprintf(stderr, "Error: Invalid key hexadecimal string\n");
+				if (free_message && file_buffer)
```
Comment (jforissier) [inline sha/host/main.c]: No need to test for file_buffer since free(NULL) is valid.
Also this pattern is repeated multiple times. I think using an error label and `goto err;` here would be better.

**Review by jforissier — COMMENTED** [review]
Looks mostly good, two comments below. Thanks.

File: sha/host/main.c (hunk 4 above)
Comment (Akshay-Belsare) [reply sha/host/main.c:124]: The <ctype.h> header has already been included.

File: sha/host/main.c [hunk 6]
```diff
@@ -648,10 +642,7 @@ int main(int argc, char *argv[])
 
 	terminate_tee_session(&ctx);
 
-	/* Cleanup file buffer if allocated */
-	if (free_message && file_buffer) {
-		free(file_buffer);
-	}
-
-	return 0;
+err:
```
Comment (jforissier) [inline sha/host/main.c]: Minor: since this path is taken also on success, I suggest renaming the label to `out:`

File: sha/host/main.c (hunk 4 above)
Comment (jforissier) [reply sha/host/main.c:124]: Indeed, how did I miss it? Sorry about that.

**Review by jforissier — COMMENTED** [review]
One minor comment. With that addressed, please add:

`Reviewed-by: Jerome Forissier <jerome.forissier@arm.com>`

Thanks!

File: sha/host/main.c (hunk 6 above)
Comment (Akshay-Belsare) [reply sha/host/main.c]: Addressed!

File: sha/host/main.c [hunk 7]
```diff
@@ -7,6 +7,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <ctype.h>
```
Comment (etienne-lms) [inline sha/host/main.c]:
```patch
+#include <ctype.h>
 #include <err.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <ctype.h>

File: sha/host/main.c [hunk 8]
```diff
@@ -17,6 +18,54 @@
...
 #define SHA_HMAC	0
 #define BASE_SHA        1
+#define CMAC            2
+
+/* Maximum sizes */
+#define MAX_KEY_SIZE    128
+#define MAX_IV_SIZE     16
+
+/* Algorithm information structure */
+typedef struct {
```
Comment (etienne-lms) [inline sha/host/main.c]: Prefer not using type definition. Could you use `struct algo_info` as type where applicable?

File: sha/host/main.c [hunk 9]
```diff
@@ -17,6 +18,54 @@
...
+	/* HMAC algorithms - require key, NO IV */
+	{"HMAC_SHA1",   TEE_ALG_HMAC_SHA1,   TA_SHA_OBJ_TYPE_HMAC_SHA1,   64,  0, true, false, SHA_HMAC},
+	{"HMAC_SHA224", TEE_ALG_HMAC_SHA224, TA_SHA_OBJ_TYPE_HMAC_SHA224, 64,  0, true, false, SHA_HMAC},
+	{"HMAC_SHA256", TEE_ALG_HMAC_SHA256, TA_SHA_OBJ_TYPE_HMAC_SHA256, 128, 0, true, false, SHA_HMAC},
+	{"HMAC_SHA384", TEE_ALG_HMAC_SHA384, TA_SHA_OBJ_TYPE_HMAC_SHA384, 128, 0, true, false, SHA_HMAC},
+	{"HMAC_SHA512", TEE_ALG_HMAC_SHA512, TA_SHA_OBJ_TYPE_HMAC_SHA512, 128, 0, true, false, SHA_HMAC},
+	{"HMAC_SHA3_224", TEE_ALG_HMAC_SHA3_224, TA_SHA_OBJ_TYPE_HMAC_SHA3_224, 128, 0, true, false, SHA_HMAC},  /* 1024 bits = 128 bytes */
+	{"HMAC_SHA3_256", TEE_ALG_HMAC_SHA3_256, TA_SHA_OBJ_TYPE_HMAC_SHA3_256, 128, 0, true, false, SHA_HMAC},  /* 1024 bits = 128 bytes */
+	{"HMAC_SHA3_384", TEE_ALG_HMAC_SHA3_384, TA_SHA_OBJ_TYPE_HMAC_SHA3_384, 128, 0, true, false, SHA_HMAC},  /* 1024 bits = 128 bytes */
+	{"HMAC_SHA3_512", TEE_ALG_HMAC_SHA3_512, TA_SHA_OBJ_TYPE_HMAC_SHA3_512, 128, 0, true, false, SHA_HMAC},  /* 1024 bits = 128 bytes */
```
Comment (etienne-lms) [inline sha/host/main.c]: Could drop the `/* 1024 bits = 128 bytes */` inline comments.

File: sha/host/main.c [hunk 10]
```diff
@@ -17,6 +18,54 @@
...
+/* Maximum sizes */
+#define MAX_KEY_SIZE    128
+#define MAX_IV_SIZE     16
+
+/* Algorithm information structure */
+typedef struct {
+	const char *name;
+	uint32_t tee_algo_id;
+	enum ta_sha_object_type obj_type;
+	size_t key_size;
```
Comment (etienne-lms) [inline sha/host/main.c]:
```suggestion
	size_t key_byte_size;
```
so that there is no ambiguity on wether its a byte or bit key size

File: sha/host/main.c [hunk 11]
```diff
@@ -203,151 +378,261 @@ static TEEC_Result compare_hmac_sha(struct test_ctx *ctx, void *message,
...
+			key_hex = argv[++i];
+		} else if (strcmp(argv[i], "--iv") == 0) {
+			if (i + 1 >= argc) {
+				fprintf(stderr, "Error: --iv requires a hexadecimal string\n");
+				print_usage(argv[0]);
+				return -1;
+			}
+			iv_hex = argv[++i];
+		} else {
+			fprintf(stderr, "Error: Unknown option '%s'\n", argv[i]);
```
Comment (etienne-lms) [inline sha/host/main.c]: Keep backward compat and assume arguments are "<string> [algo]" when no option starting with `--` is passed? There may be existing test bench that use the legacy command arguments.

File: sha/host/main.c [hunk 12]
```diff
@@ -203,151 +378,261 @@ static TEEC_Result compare_hmac_sha(struct test_ctx *ctx, void *message,
...
+	int rc = 0;
+
+	/* User-provided or default key/IV */
+	uint8_t user_key[MAX_KEY_SIZE];
+	size_t user_key_len = 0;
+	uint8_t user_iv[MAX_IV_SIZE];
+	size_t user_iv_len = 0;
+
+	/* Parse command-line arguments */
+	for (int i = 1; i < argc; i++) {
```
Comment (etienne-lms) [inline sha/host/main.c:403]: Could you move argument parsing in a local helper function, to keep `main()` simple to read?
Not a blocking issue, could be done in a dedicated P-R.

File: sha/host/main.c [hunk 13]
```diff
@@ -203,151 +378,261 @@ static TEEC_Result compare_hmac_sha(struct test_ctx *ctx, void *message,
...
+			user_iv_len = algo_info->iv_size;
+		}
 	}
 
 	printf("Prepare session with the TA\n");
 	prepare_tee_session(&ctx);
 
-	if (algo_type != SHA_HMAC) {
-		/* SHA digest */
+	if (algo_info->algo_type == BASE_SHA) {
```
Comment (etienne-lms) [inline sha/host/main.c]: Could you use `switch/case` instructions here?

File: sha/host/main.c [hunk 14]
```diff
@@ -50,18 +99,144 @@ static void terminate_tee_session(struct test_ctx *ctx)
...
+	*out_len = byte_len;
+	return 0;
+}
+
+/*
+ * Get algorithm info by name
+ */
+static const algo_info_t *get_algo_info(const char *name)
+{
+	for (int i = 0; algo_table[i].name != NULL; i++) {
```
Comment (etienne-lms) [inline sha/host/main.c]: For consistency:
```suggestion
	int i = 0;

	for (i = 0; algo_table[i].name != NULL; i++) {

File: sha/host/main.c [hunk 15]
```diff
@@ -50,18 +99,144 @@ static void terminate_tee_session(struct test_ctx *ctx)
...
+	if (!f) {
+		fprintf(stderr, "Error: Cannot open file '%s'\n", filepath);
+		return -1;
+	}
+
+	/* Get file size */
+	if (fseek(f, 0, SEEK_END) != 0) {
+		fprintf(stderr, "Error: Cannot seek to end of file '%s'\n", filepath);
+		fclose(f);
+		return -1;
```
Comment (etienne-lms) [inline sha/host/main.c]: Suggestion (feel free to discard):
```patch
+	int ret = -1;
 
 	...
 
 	if (fseek(f, 0, SEEK_END) != 0) {
 		fprintf(stderr, "Error: Cannot seek to end of file '%s'\n", filepath);
+		goto out;
 	}
 
 	...
 
 	*buffer = data;
 	*size = (size_t)file_size;
+	ret = 0;
 
+out:
 	fclose(f);
+	if (ret)
+		free(data);

	return ret;
}

**Review by etienne-lms — COMMENTED** [review]
No blocking issue but the `typedef`. I'd prefer we don't start defining types when no strictly needed.
Also wondering about the backward compatibility on command arguments. That said, maybe `optee_example_sha` is recent enough so that there won't be compat issues.

File: sha/host/main.c (hunk 12 above)
Comment (amey-raghatate) [reply sha/host/main.c:403]: Agreed, this is a good suggestion. We will address this in a dedicated follow-up PR to keep this PR focused.
